gematria/datasets/process_and_filter_bbs.cc

```diff
@@ -0,0 +1,166 @@
+    "filter-memory-accessing-blocks",
+    cl::desc("Whether or not to filter out blocks that access memory"),
+    cl::init(false), cl::cat(ProcessFilterCat));
+
+static cl::opt<unsigned> ReportProgressEvery(
+    "report-progress-every",
+    cl::desc("The interval at which to report progress in blocks"),
+    cl::init(std::numeric_limits<unsigned>::max()), cl::cat(ProcessFilterCat));
+
+static cl::opt<unsigned> MaxThreadCount(
+    "thread-count",
+    cl::desc("The maximum number of threads to use to process BBs"),
+    cl::init(llvm::thread::hardware_concurrency()), cl::cat(ProcessFilterCat));
+
+static cl::opt<unsigned> BatchSize(
+    "batch-size",
+    cl::desc("The number of blocks to include in a batch to be processed by a "
+             "single thread"),
+    cl::init(1000), cl::cat(ProcessFilterCat));
+
+Expected<std::string> ProcessBasicBlock(
+    const std::string &BasicBlock,
+    const gematria::LlvmArchitectureSupport &LLVMSupport,
+    MCInstPrinter &MachineInstructionPrinter, const StringRef FileName) {
+  // TODO(boomanaiden154): Update this to use llvm::Expected once
+  // gematria::ParseHex is refactored to return llvm::Expected.
+  auto MachineCodeHex = gematria::ParseHexString(BasicBlock);
+  if (!MachineCodeHex.has_value()) {
+    return createFileError(
+        FileName,
+        make_error<StringError>(
+            llvm::Twine("Could not parse: '").concat(BasicBlock).concat("'"),
+            std::make_error_code(std::errc::invalid_argument)));
+  }
+
+  Expected<std::vector<gematria::DisassembledInstruction>>
+      DisassembledInstructionsOrErr = gematria::DisassembleAllInstructions(
+          LLVMSupport.mc_disassembler(), LLVMSupport.mc_instr_info(),
+          LLVMSupport.mc_register_info(), LLVMSupport.mc_subtarget_info(),
+          MachineInstructionPrinter, 0, *MachineCodeHex);
+
+  if (!DisassembledInstructionsOrErr)
+    return createFileError(FileName, DisassembledInstructionsOrErr.takeError());
+
+  std::string OutputBlock;
+
+  for (const gematria::DisassembledInstruction &Instruction :
+       *DisassembledInstructionsOrErr) {
+    MCInstrDesc InstDesc =
+        LLVMSupport.mc_instr_info().get(Instruction.mc_inst.getOpcode());
+    if (Instruction.mc_inst.getOpcode() == X86::SYSCALL) continue;
+    if (InstDesc.isReturn() || InstDesc.isCall() || InstDesc.isBranch())
+      continue;
+    if (FilterMemoryAccessingBlocks &&
+        (InstDesc.mayLoad() || InstDesc.mayStore()))
+      continue;
+    OutputBlock += toHex(Instruction.machine_code);
+  }
+
+  return OutputBlock;
+}
+
+int main(int Argc, char **Argv) {
+  cl::ParseCommandLineOptions(Argc, Argv, "process_and_filter_bbs");
+
+  ExitOnError ExitOnErr("process_and_filter_bbs error: ");
+
+  const std::unique_ptr<gematria::LlvmArchitectureSupport> LLVMSupport =
+      gematria::LlvmArchitectureSupport::X86_64();
+
+  std::unique_ptr<MCInstPrinter> MachineInstructionPrinter =
+      LLVMSupport->CreateMCInstPrinter(0);
```

Comment:
When we run multiple threads, is it safe to use the same instruction printer from all of them at the same time?